### PR TITLE
Allow for skipping navigation link creation

### LINKED ISSUE #15960: Code of Conduct, Privacy Policy, and Terms of Use Nav Links Keep Coming Back After Being Deleted
**Describe the bug**

A number of non-DEV Forems are hoping to remove the default navigation links for Code of Conduct, Privacy Policy, and Terms of Use. They are able to successfully delete these links, but the links eventually come back. We're thinking this is likely happening when a Creator updates their cloud instance to the latest version of Forem.

**To Reproduce**

Here's [a video](https://www.loom.com/share/362b6914a7fd4eddb32f82355a481791) that shows how to remove the nav links.

Here's a more recent screenshot of jess.forem.lol (our Forem for testing) that shows the nav links have returned:

Screen Shot 2022-01-05 at 9 44 23 AM

**Expected behavior**

Deleted nav links should be removed permanently unless they are added back. Updating the Forem should not revert any removed nav links.

### PR DESCRIPTION
**Pull request by jeremyf**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [x] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This commit provides a possible solution for preventing the re-creation
of a navigation link deleted by a Forem creator.

What I need is a discussion around the life-cycle of the application
installation and updates.

In particular, does this provide a robust enough mechanism for resolving
the issue at hand?

_Note: it pains me to ask about a user's role, but this is provided as a
point of discussion and possible implementationi to address the
underlying issue.  But I'm referencing a constant in the Rake task so
hopefully that will help future refactors.  Also, it's one reason I
needed to remove the `private_constant` declaration._

If we accept this code change, a future task is to document the ENV and
behavior.

### Further considerations

- How might we refine this to not be as "role" reliant?
- Could we have a Site::Setting that we enable/disable
  regarding the navigation links?

## Related Tickets & Documents

Closes #15960

## QA Instructions, Screenshots, Recordings

Here's the steps I took.  Note the guard clause means there are several ways to create or skip creating.

- Start from an empty database
- Run setup
- Verify Navigation Link exists
- Delete Navigation Link
- Run setup again
- Verify Navigation Link exists (because we don't have a user)
- Run seeds
- Delete Navigation Link
- Run setup again
- Verify Navigation Link exists (because we don't have a user)

```shell
$ cd ./path/to/forem/repo

$ rails db:drop db:create db:schema:load

$ bin/setup

$ bin/rails runner "puts NavigationLink.where(url: '/readinglist').exists?"
  => true

$ bin/rails runner "NavigationLink.where(url: '/readinglist').delete_all"

$ bin/rails runner "puts NavigationLink.where(url: '/readinglist').exists?"
  => false

$ bin/setup

$ bin/rails runner "puts NavigationLink.where(url: '/readinglist').exists?"
  => true

$ bin/rails db:seed

$ bin/rails runner "NavigationLink.where(url: '/readinglist').delete_all"

$ bin/rails runner "puts NavigationLink.where(url: '/readinglist').exists?"
  => false

$ bin/setup

$ bin/rails runner "puts NavigationLink.where(url: '/readinglist').exists?"c
  => false
```

## Added/updated tests?

- [x] No, and this is why: We don't have tests for these kind of things; they are a bit hard and slow to run.

## [Forem core team only] How will this change be communicated?

- [ ] This PR changes the Forem platform and our documentation needs to be
      updated. I have filled out the
      [Changes Requested](https://github.com/forem/admin-docs/issues/new?assignees=&labels=&template=changes_requested.md)
      issue template so Community Success can help update the Admin Docs
      appropriately.
- [x] I will share this change internally with the appropriate teams

_Note: I have not updated the documentation as this is pending discussion._

### CODE REVIEW

File: lib/tasks/add_navigation_links.rake [hunk 1]
```diff
@@ -22,48 +22,93 @@ namespace :navigation_links do
...
+  def perform_create_of_navigation_links?
+    # Someone really wants this
+    return true if ApplicationConfig["CREATE_NAVIGATION_LINKS"]
+
+    # This logic echoes the InternalPolicy behavior which is used in the
+    # Admin::APpliciationController.
+    return false if User.with_any_role(*Authorizer::RoleBasedQueries::ANY_ADMIN_ROLES).any?
+
+    # May as well provide a way for them to skip
+    return false if ApplicationConfig["SKIP_CREATE_NAVIGATION_LINKS"]
```
Comment (citizen428) [inline lib/tasks/add_navigation_links.rake:34]: Are we sure about that? I think some things assume these links will be there, so maybe YAGNI until we tackle this more comprehesively?

File: lib/tasks/add_navigation_links.rake [hunk 2]
```diff
@@ -22,48 +22,93 @@ namespace :navigation_links do
...
+      #            desired.  I did not add descriptions so those tasks will not show up in the task
+      #            list.
+      Rake::Task["navigation_links:find_or_create:readinglist"].invoke
+      Rake::Task["navigation_links:find_or_create:contact"].invoke
+      Rake::Task["navigation_links:find_or_create:code_of_conduct"].invoke
+      Rake::Task["navigation_links:find_or_create:privacy"].invoke
+      Rake::Task["navigation_links:find_or_create:terms"].invoke
+    else
+      # Adding just a bit of logging
+      puts "Skipping creation of navigation links"
```
Comment (citizen428) [inline lib/tasks/add_navigation_links.rake:53]: Usually, Rubocop complains when you try to `puts` for logging purposes.
```suggestion
      Rails.logger.info "Skipping creation of navigation links"
```

File: lib/tasks/add_navigation_links.rake [hunk 3]
```diff
@@ -22,48 +22,93 @@ namespace :navigation_links do
   heart_icon = twemoji_path("heart.svg")
   rainbowdev = image_path("rainbowdev.svg")
 
+  def perform_create_of_navigation_links?
+    # Someone really wants this
+    return true if ApplicationConfig["CREATE_NAVIGATION_LINKS"]
+
+    # This logic echoes the InternalPolicy behavior which is used in the
+    # Admin::APpliciationController.
```
Comment (citizen428) [inline lib/tasks/add_navigation_links.rake:30]:
```suggestion
    # Admin::AppliciationController.
```

File: yarn.lock [hunk 4]
```diff
@@ -2306,6 +2306,40 @@
     global "^4.4.0"
     regenerator-runtime "^0.13.7"
 
+"@storybook/addons@6.4.9":
```
Comment (citizen428) [inline yarn.lock:2309]: I don't think there should be any `yarn.lock` changes in this PR.

**Review by citizen428 — APPROVED** [review]
I think this works as a stop-gap measure to solve the current problem but in the mid to long-term we probably want a cleaner defined lifecycle so it gets easier to differentiate between first install and later deployments.

**Comment by jeremyf** [comment]
The `yarn.lock` changes appear to be a necessary issue for Buildkite (see https://buildkite.com/forem/build-containers/builds/10583#c5224173-dbec-4bcc-883f-be441a9bbd0a)

**Review by maestromac — APPROVED** [review]

**Pull request merged**